### PR TITLE
update cartservice

### PR DESCRIPTION
**Pull request by mathieu-benoit**
- Size of the image: `200MB` --> `50.3MB`
- Update `Grpc.*` nuget packages to `2.34.0`
- Simplify the `Program.Start` mechanism
- [Single-File Executables in .NET](https://levelup.gitconnected.com/single-file-executables-in-net-core-3-1-and-the-quest-for-a-sub-50mb-docker-container-f44cb1274121)
- [App Trimming in .NET 5](https://devblogs.microsoft.com/dotnet/app-trimming-in-net-5/)
- Misc cleanups

### CODE REVIEW

File: src/cartservice/.dockerignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+**/*.sh
+**/*.bat
```
Comment (simonz130) [inline src/cartservice/.dockerignore:2]: why these files are ignored?

File: src/cartservice/.dockerignore [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+tests/
+**/bin/
+**/obj/
+**/out/
+Dockerfile*
+docker-compose*
```
Comment (simonz130) [inline src/cartservice/.dockerignore:8]: similarly, here?

File: src/cartservice/Program.cs [hunk 3]
```diff
@@ -12,116 +12,15 @@
...
-using System.Runtime.CompilerServices;
+using cartservice;
 
-[assembly: InternalsVisibleTo("cartservice.tests")]
-namespace cartservice
-{
-    public class Program
-    {
-        const string REDIS_ADDRESS = "REDIS_ADDR";
+CreateHostBuilder(args).Build().Run();
```
Comment (simonz130) [inline src/cartservice/Program.cs:19]: I'm not sure I understand how this even compiles - it looks like you are removing the namespace and class and this method is not contained anywhere. Is it a glitch in the comparison?

File: kubernetes-manifests/cartservice.yaml [hunk 4]
```diff
@@ -38,8 +38,6 @@ spec:
         env:
         - name: REDIS_ADDR
           value: "redis-cart:6379"
-        - name: ASPNETCORE_URLS
-          value: "http://0.0.0.0:7070"
         resources:
```
Comment (simonz130) [inline kubernetes-manifests/cartservice.yaml:41]: Did you test it deployed on kubernetes?

File: tests/cartservice/cartservice.tests.csproj [hunk 5]
```diff
@@ -1,19 +1,19 @@
...
-    <PackageReference Include="Grpc.Net.Client" Version="2.32.0" />
-    <PackageReference Include="Grpc.Tools" Version="2.23.0" />
-    <PackageReference Include="Google.Protobuf" Version="3.12.3" />
-    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.7.0" />
-    <PackageReference Include="xunit" Version="2.3.1" />
-    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
-    <DotNetCliToolReference Include="dotnet-xunit" Version="2.3.1" />
+    <PackageReference Include="Grpc.Net.Client" Version="2.34.0" />
+    <PackageReference Include="Grpc.Tools" Version="2.34.0" />
+    <PackageReference Include="Google.Protobuf" Version="3.14.0" />
```
Comment (simonz130) [inline tests/cartservice/cartservice.tests.csproj:12]: Did you test that tests for cart service pass with your code?

File: src/cartservice/docker-compose.yaml [hunk 6]
```diff
@@ -23,6 +23,5 @@ services:
       - "7070:7070"
     environment:
       REDIS_ADDR: "redis:6379"
-      ASPNETCORE_URLS: "http://0.0.0.0:7070"
```
Comment (simonz130) [inline src/cartservice/docker-compose.yaml:26]: why was this removed?

File: src/cartservice/Services/CartServiceController.cs [hunk 7]
```diff
@@ -45,11 +45,6 @@ public async override Task<Empty> EmptyCart(EmptyCartRequest request, ServerCall
         {
             await _cartStore.EmptyCartAsync(request.UserId);
             return Empty;
-        } 
-        public async void Start()
-        {
-            Console.WriteLine("Starting cart service...");
-            await _cartStore.InitializeAsync();
```
Comment (simonz130) [inline src/cartservice/Services/CartServiceController.cs:52]: Where do you initialize cartStore?

**Review by simonz130 — CHANGES_REQUESTED** [review]

File: src/cartservice/Program.cs (hunk 3 above)
Comment (mathieu-benoit) [reply src/cartservice/Program.cs:19]: Good question, that's one of the improvements with C# 9 described here: https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/#top-level-programs.

File: src/cartservice/Services/CartServiceController.cs (hunk 7 above)
Comment (mathieu-benoit) [reply src/cartservice/Services/CartServiceController.cs:52]: This method was not used, the initialization is done in `Startup.cs`: https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/blob/mathieu-benoit/update-cartservice/src/cartservice/Startup.cs#L52

File: src/cartservice/.dockerignore (hunk 2 above)
Comment (mathieu-benoit) [reply src/cartservice/.dockerignore:8]: Those files should not go the container itself, so to avoid any `COPY` command in the `Dockerfile` bringing them, we could add their exclusion in the `.dockerignore` file as a best practice.

File: src/cartservice/.dockerignore (hunk 1 above)
Comment (mathieu-benoit) [reply src/cartservice/.dockerignore:2]: Those type of files should not go the container itself (at least in the context of the `cartservice`), so to avoid any `COPY` command in the `Dockerfile` bringing them, we could add their exclusion in the `.dockerignore` file as a best practice.

File: src/cartservice/Dockerfile [hunk 8]
```diff
@@ -12,34 +12,18 @@
...
+FROM mcr.microsoft.com/dotnet/sdk:5.0.101 as builder
 WORKDIR /app
-COPY *.csproj .
-
-# Workaround for https://github.com/grpc/grpc/issues/18428
-# Also see https://github.com/sgerrand/alpine-pkg-glibc
-RUN apk update && apk --no-cache add ca-certificates wget
-RUN wget -q -O /etc/apk/keys/sgerrand.rsa.pub https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub
-RUN wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.30-r0/glibc-2.30-r0.apk
-RUN apk add glibc-2.30-r0.apk
```
[the remark this answers is not in the document]
Comment (mathieu-benoit) [reply src/cartservice/Dockerfile:24]: I went through those links (workaround, etc.) mentioned, and now that we got the managed gRPC implementation with your PR https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/pull/494, we shouldn't need this anymore (`Grpc.Tools` not used anymore for that part).
With that said, I don't see why on MacOS this shouldn't work, could you confirm that's needed for MacOS (I don't see any reference of this in the workaround link)? During the `docker build` or `docker run` part? I don't have access to a MacOS to test it unfortunately.
_Note: if you have an issue on MacOS, I think I know what could be the issue, but first let's checked the above points before going that path._

File: src/cartservice/docker-compose.yaml (hunk 6 above)
Comment (mathieu-benoit) [reply src/cartservice/docker-compose.yaml:26]: This environment variable is now part of the `Dockerfile`, so no need to add it here.

File: src/cartservice/.dockerignore (hunk 1 above)
Comment (simonz130) [reply src/cartservice/.dockerignore:2]: Agree on the bat files for the container but not necessarily for sh files. Let's not ignore *.sh - in case we might want to wrap the microservice executable in a script

File: src/cartservice/Program.cs (hunk 3 above)
Comment (simonz130) [reply src/cartservice/Program.cs:19]: Let's revert to have a main method. .NET documentation/tutorials throughout the internet haven't picked up all the new features from the .NET team and many of the customers might find this confusing.

File: kubernetes-manifests/cartservice.yaml (hunk 4 above)
Comment (mathieu-benoit) [reply kubernetes-manifests/cartservice.yaml:41]: Yes I did.

Comment (simonz130) [reply kubernetes-manifests/cartservice.yaml:41]: I prefer to set environment variables for kubernetes clusters in kubernetes manifest instead of docker file. If you then decide to use buildbacks for container building, you might introduce errors in configuration as a result of handling environment-related stuff in dockerfile instead of kubernetes yaml

Comment (mathieu-benoit) [reply kubernetes-manifests/cartservice.yaml:41]: Having it in the `Dockerfile` allows to have centralized way to get thei `cartservice` app working on any circunstances (`Docker`, Docker-compose`, `Kubernetes`, etc.). It's not a Kubernetes thing, for example today this command `docker run cartservice` won't work unless `-e ASPNETCORE_URLS="http://0.0.0.0:7070"` is passed.

File: src/cartservice/.dockerignore (hunk 1 above)
Comment (mathieu-benoit) [reply src/cartservice/.dockerignore]: Done.

File: src/cartservice/Program.cs (hunk 3 above)
Comment (mathieu-benoit) [reply src/cartservice/Program.cs]: Done.

File: kubernetes-manifests/cartservice.yaml (hunk 4 above)
Comment (mathieu-benoit) [reply kubernetes-manifests/cartservice.yaml]: Done.

File: src/cartservice/Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (mathieu-benoit) [reply src/cartservice/Dockerfile:24]: Quick update here, I'm able to reproduce the issue explained with the workaround (on Linux) if the `builder` image is `alpine` which is not the case with my update which doesn't have the issue and don't need that workaround: `mcr.microsoft.com/dotnet/sdk:5.0.101`.

File: tests/cartservice/cartservice.tests.csproj (hunk 5 above)
Comment (mathieu-benoit) [reply tests/cartservice/cartservice.tests.csproj:12]: I didn't, but I'm on it and will report back once done. In the meantime, I have hard time to get the `tests/carservice/tests.yaml` working, is it working on `master`?

Comment (mathieu-benoit) [reply tests/cartservice/cartservice.tests.csproj:11]: I was able to successfully run the tests, but I needed to apply a fix https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/pull/612/commits/7c8c7479e717e8953552a23fdbce95c7a31d1345.

File: src/cartservice/Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (simonz130) [reply src/cartservice/Dockerfile:24]: Awesome. Then sounds like we can remove it!

**Review by simonz130 — APPROVED** [review]
